portalnet/src/gossip.rs

```diff
@@ -243,84 +214,105 @@ pub async fn trace_propagate_gossip_cross_thread<TContentKey: OverlayContentKey>
+/// associated with it will be empty.
+fn batch_interested_enrs<TMetric: Metric>(
+    content_ids: &[&[u8; 32]],
+    kbuckets: Arc<RwLock<KBucketsTable<NodeId, Node>>>,
+) -> HashMap<[u8; 32], Vec<Enr>> {
+    let mut result = content_ids
+        .iter()
+        .map(|content_id| (**content_id, vec![]))
+        .collect::<HashMap<_, _>>();
+    for entry in kbuckets.write().iter() {
+        // Skip non-connected nodes
+        if !entry.status.is_connected() {
+            continue;
+        }
+        let node = entry.node;
+        for content_id in content_ids {
+            let distance = TMetric::distance(content_id, &node.key.preimage().raw());
+            if distance <= node.value.data_radius {
+                result
+                    .entry(**content_id)
+                    .or_default()
+                    .push(node.value.enr());
+            }
+        }
+    }
+    result
 }
 
 const NUM_CLOSEST_NODES: usize = 4;
 const NUM_FARTHER_NODES: usize = 4;
-/// Selects gossip recipients from a vec of sorted interested ENRs.
-/// Returned vec is a concatenation of, at most:
-/// 1. First `NUM_CLOSEST_NODES` elements of `interested_sorted_enrs`.
-/// 2. `NUM_FARTHER_NODES` elements randomly selected from
-///    `interested_sorted_enrs[NUM_CLOSEST_NODES..]`
-fn select_gossip_recipients(interested_sorted_enrs: Vec<Enr>) -> Vec<Enr> {
-    let mut gossip_recipients: Vec<Enr> = vec![];
 
-    // Get first n closest nodes
-    gossip_recipients.extend(
-        interested_sorted_enrs
-            .clone()
-            .into_iter()
-            .take(NUM_CLOSEST_NODES),
-    );
-    if interested_sorted_enrs.len() > NUM_CLOSEST_NODES {
-        let farther_enrs = interested_sorted_enrs[NUM_CLOSEST_NODES..].to_vec();
-        // Get random non-close ENRs to gossip to.
-        let random_farther_enrs = select_random_enrs(NUM_FARTHER_NODES, farther_enrs);
-        gossip_recipients.extend(random_farther_enrs);
+/// Selects gossip recipients from a vec of interested ENRs.
+///
+/// If number of ENRs is at most `NUM_CLOSEST_NODES + NUM_FARTHER_NODES`, then all are returned.
+/// Otherwise, ENRs are sorted by distance from `content_id` and then:
+///
+/// 1. Closest `NUM_CLOSEST_NODES` ENRs are selected
+/// 2. Random `NUM_FARTHER_NODES` ENRs are selected from the rest
+fn select_gossip_recipients<TMetric: Metric>(
+    content_id: &[u8; 32],
+    mut enrs: Vec<Enr>,
+) -> Vec<Enr> {
+    // Check if we need to do any selection
+    if enrs.len() <= NUM_CLOSEST_NODES + NUM_FARTHER_NODES {
+        return enrs;
     }
-    gossip_recipients
+
+    // Sort enrs by distance
+    enrs.sort_by_cached_key(|enr| TMetric::distance(content_id, &enr.node_id().raw()));
+
+    // Split of at NUM_CLOSEST_NODES
+    let farther_enrs = enrs.split_off(NUM_CLOSEST_NODES);
+
+    // Select random NUM_FARTHER_NODES
+    let random_farther_enrs = farther_enrs
+        .into_iter()
+        .choose_multiple(&mut rand::thread_rng(), NUM_FARTHER_NODES);
```

Comment:
I also think it's of little practical importance, but since we started debate, let's go deeper :)

I see 3 ways to do it:
```rust
// 1. Shuffle: clean, very readable, no cloning. BAD: O(farther_enrs.len())
farther_enrs.shuffle(&mut rand::thread_rng());
farther_enrs.truncate(NUM_FARTHER_NODES);
enrs.extend(farther_enrs);

// 2. Partial shuffle: clean, O(NUM_FARTHER_NODES), BAD: we have to clone ENRs
let (selected_farther_enrs, _other_farther_enrs) =
   farther_enrs.partial_shuffle(&mut rand::thread_rng(), NUM_FARTHER_NODES);
enrs.extend_from_slice(selected_farther_enrs);

// 3. Manual: simple, O(NUM_FARTHER_NODES), no cloning. BAD: not as clean, very manual
let mut rng = rand::thread_rng();
for _ in 0..NUM_FARTHER_NODES {
    let enr = farther_enrs.swap_remove(rng.gen_range(0..farther_enrs.len()));
    enrs.push(enr);
}
```

First one is the cleanest, but worse performance (question is how much it matters).
Second one has good performance, but we have to clone ENRs (again, it's only 4 of them, does it matter?)
Last one, that I ended up going with, has both best performance and no cloning. The downside is that I don't find it as "clean" as the other two, but I think it's simple and clean enough.

I don't have strong preferences to any of the them, so if you do, I'm willing to change it.

Reply:
Hah, a good bit of performance fun. Good catch on the cloning. LGTM!